src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,486 @@
+use serde::ser::{Serialize};
+use serde_json::ser::{Formatter, CharEscape};
+use std::io::Write;
+use log::{debug, error};
+
+struct JSONFormatter {}
+
+impl Formatter for JSONFormatter {
+    fn write_f64<W: ?Sized>(&mut self, writer: &mut W, value: f64) -> Result<(), std::io::Error> where
+        W: Write,
+    {
+        writer.write(&format_number(value, 10.0_f64.powf(21.0), 10.0_f64.powf(-6.0)).into_bytes())?;
+        Ok(())
+    }
+
+    fn write_char_escape<W: ?Sized>(&mut self, writer: &mut W, char_escape: CharEscape) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        match char_escape {
+            CharEscape::Quote => {
+                writer.write(&"\"".to_owned().into_bytes())?; 
+            },
+            CharEscape::ReverseSolidus => {
+                writer.write(&"\\\\".to_owned().into_bytes())?;
+            },
+            CharEscape::LineFeed => {
+                writer.write(&"\n".to_owned().into_bytes())?;
+            },
+            CharEscape::Tab => {
+                writer.write(&"\t".to_owned().into_bytes())?;
+            },
+            CharEscape::CarriageReturn => {
+                writer.write(&"\r".to_owned().into_bytes())?;
+            },
+            _ => { println!("Other char escape"); }
```

Comment:
Print is definitely wrong.  Is this an error?  Something to silently ignore?  Something to log?

Reply:
other types in CharEscape like backspace and formfeed are not recognized by Rust

Reply:
Backspace, Formfeed char escapes are not recognized by Rust so there is nothing to do for them

Reply:
If there are a small number of them, then mach them explicitly and do nothing.  If there are a large number, then `_` might be best -- just say what you told me in a comment.

Reply:
yes I have fixed it